docs: synchronization and paralellism docs improvements (#13393)
argoproj/argo-workflows: docs/parallelism.md

# Limiting parallelism

You can restrict the number of simultaneous workflow executions.

## Controller level

You can limit the total number of workflows that can execute at any one time in the [workflow controller ConfigMap](./workflow-controller-configmap.yaml).

```yaml
data:
  parallelism: "10"
```

> You can also limit the number of workflows that can execute in a single namespace.

Reviewer: ```suggestion
You can also limit the total number of parallel workflow executions in a single namespace:
```
same as above

```yaml
data:
  namespaceParallelism: "4"
```

Workflows that are executing but restricted from running more nodes due to the other mechanisms will still count towards the parallelism limits.

> ### Priority

Reviewer: Should this be called "Queueing" to match the "Synchronization" page?

```suggestion
### Queueing
```

Reviewer: I also wonder if we should just link to the "Synchronization" page, as, if I'm not mistaken, it's effectively identical here

Author: The queue is an implementation detail from the users perspective. All they care about is that priority exists and they can use it to prioritise. They don't need to know this is implemented by queueing.

Reviewer: I made that comment before your change to link them, so let me re-read the new variant

> Workflows can have a `priority` set in their specification.

Reviewer: ```suggestion
You can set a `priority` on Workflows.
```

active voice, same as above

Author: Same problem as above.

Author: I've added a description of how to use this, and the missing information of default priority.

> Workflows with a higher priority number that have not started due to controller level parallelism will be started before lower priority workflows.

Reviewer: ```suggestion
Workflows that have not started yet will be queued: Workflows with higher priority numbers will start before lower priority ones.
```

simplify.

although this feels identical to the "Queuing" section in the "Synchornization" page as I mentioned above

Author: This is a different queue, unrelated to any synchronization based queues.

You need to pass through this queue before you can enter any synchronization based queue, and this is important to understand.

In the Synchronization section the queue is about "What happens when a mutex isn't available". At the moment you get no option, you enter a queue. We may in future have alternative options to skip this section of the workflow, or wait in the queue for some limited time.

It's probable that people can guess they would enter a queue, and here the details they cannot guess at are the ordering of the queue. It's not a FIFO as you may guess as a simple implementation, it's priority then workflow creation time.

Author: Your rewrite here doesn't explain why workflows may not be started, so I've adopted some of your rewrite.

## Synchronization

> You can use [mutexes, semaphores and parallelism](./synchronization.md) to control the parallel execution of sections of a workflow.

Reviewer: ```suggestion
You can also use [mutexes, semaphores, and parallelism](synchronization.md) to control the parallel execution of workflows and templates.
```

- Oxford comma pl0x
- remove `./` in link, same as above
- it's not just "sections of a workflow", since it is a superset, so I added "also" and used "workflows or templates" as that page does

# Limiting parallelism

You can restrict the number of parallel workflow executions.

## Controller level

You can limit the total number of parallel workflow executions in the [workflow controller ConfigMap](workflow-controller-configmap.yaml):

```yaml
data:
  parallelism: "10"
```

You can also limit the total number of parallel workflow executions in a single namespace:

```yaml
data:
  namespaceParallelism: "4"
```

!!! Note
    Workflows that are executing but restricted from running more nodes due to other mechanisms will still count toward parallelism limits.

### Priority

You can set a priority on Workflows using the `priority` field in the specification

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: priority-
spec:
  priority: 3
  ...

> ````

Reviewer: Fixed that myself in [this commit](https://github.com/argoproj/argo-workflows/pull/13393/commits/6e78e889344df95bdb9b1bdc79e753df9ecb7df1)

# Limiting parallelism

> You can restrict the number of workflows being executed at any time using these mechanisms.

Reviewer: ```suggestion
You can restrict the number of workflow executions.
```

- simplify

Author: I think the extra words here are important.
"Restrict the number of workflow executions." is not enough detail. Total number of executions ever?

I'll go with "You can restrict the number of simultaneous workflow executions."

Reviewer: That was the exact word I was thinking to add 🙂👍

## Controller level

You can limit the total number of workflows that can execute at any one time in the [workflow controller ConfigMap](./workflow-controller-configmap.yaml).

```yaml
data:
  parallelism: "10"
```

You can also limit the number of workflows that can execute in a single namespace.

```yaml
data:
  namespaceParallelism: "4"
```

> Workflows that are executing but restricted from running more nodes due to the other mechanisms will still count towards the parallelism limits.

Reviewer: hmmm there is a WIP PR to change this: #12799. if we consider that as "buggy" behavior (as the issue and PR do), then we probably shouldn't document it as it is unintended

Author: I don't only mean that mechanism. Suspended workflows will also do this.

I think the original intention would have been not to do what #12799 is trying to fix, because it's simpler. (The current fix there violates parallelism. An aware user could deliberately violate parallelism using it to get their work done.)

I'd like the current documentation to document current behaviour, so I'd like to keep this in.

Reviewer: Yea I think there's a few other cases that could all be considered "buggy". The primary difference though, is if it's documented, people won't file bug reports or fix it necessarily -- it's expected behavior. If we link to an existing bug report, or otherwise state that this behavior may be optimized in the future, I think that could be ok as a compromise.

Regarding #12799, please do review there if you have an opinion on the approach! The initial approach was to requeue, which was a lot simpler, the current one is a bit different and I honestly have yet to think through correctness on it

### Priority

Workflows can have a `priority` set in their specification.

Workflows with a higher priority number that have not started due to controller level parallelism will be started before lower priority workflows.

## Workflow level

You can restrict parallelism within a workflow using `parallelism` within a workflow or template.
This only restricts total concurrent executions of steps or tasks within the same workflow.

> Examples

Reviewer: ```suggestion
Examples:
```

missing colon

> 1 [`parallelism-limit.yaml`](https://github.com/argoproj/argo-workflows/blob/main/examples/parallelism-limit.yaml) restricts the parallelism of a [loop](./walk-through/loops.md)
> 1 [`parallelism-nested.yaml`](https://github.com/argoproj/argo-workflows/blob/main/examples/parallelism-nested.yaml) restricts the parallelism of a nested loop
> 1 [`parallelism-nested-dag.yaml`](https://github.com/argoproj/argo-workflows/blob/main/examples/parallelism-nested-dag.yaml) restricts the number of dag tasks that can be run at any one time
> 1 [`parallelism-nested-workflow.yaml`](https://github.com/argoproj/argo-workflows/blob/main/examples/parallelism-nested-workflow.yaml) shows how parallelism is inherited by children
> 1 [`parallelism-template-limit.yaml`](https://github.com/argoproj/argo-workflows/blob/main/examples/parallelism-template-limit.yaml) shows how parallelism of looped templates is also restricted

Reviewer: ```suggestion
1. [`parallelism-limit.yaml`](https://github.com/argoproj/argo-workflows/blob/main/examples/parallelism-limit.yaml) restricts the parallelism of a [loop](./walk-through/loops.md)
1. [`parallelism-nested.yaml`](https://github.com/argoproj/argo-workflows/blob/main/examples/parallelism-nested.yaml) restricts the parallelism of a nested loop
1. [`parallelism-nested-dag.yaml`](https://github.com/argoproj/argo-workflows/blob/main/examples/parallelism-nested-dag.yaml) restricts the number of dag tasks that can be run at any one time
1. [`parallelism-nested-workflow.yaml`](https://github.com/argoproj/argo-workflows/blob/main/examples/parallelism-nested-workflow.yaml) shows how parallelism is inherited by children
1. [`parallelism-template-limit.yaml`](https://github.com/argoproj/argo-workflows/blob/main/examples/parallelism-template-limit.yaml) shows how parallelism of looped templates is also restricted
```

Dots in front of the `1` were missing, which caused invalid formatting
